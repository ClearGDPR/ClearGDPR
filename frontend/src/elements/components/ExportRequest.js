import React from 'react';
import FileSaver from 'file-saver';

class ExportRequest extends React.PureComponent {
  state = {
    processing: false,
    success: false,
    error: null
  };

  eraseData() {
    this.setState({ processing: true });
    setTimeout(() => {
      const cgToken = localStorage.getItem('cgToken');
      window.cg.setAccessToken(cgToken);
      window.cg.Subject.accessData()
        .then(data => {
          const blob = new Blob([JSON.stringify(data)], { type: 'text/plain;charset=utf-8' });
          FileSaver.saveAs(blob, 'personal_data.json');
          this.setState({
            success: true,
            processing: false,
            err: null
          });
        })
        .catch(err => {
          this.setState({
            success: false,
            processing: false,
            error: err
          });
        });
    }, 1000);
  }

  render() {
    const { processing } = this.state;

    return (
      <button className="button is-primary" onClick={() => this.eraseData()} disabled={processing}>
        {!processing ? (
          <React.Fragment>{this.props.options.label}</React.Fragment>
        ) : (
          <React.Fragment>Downloading...</React.Fragment>
        )}
      </button>
    );
  }
}

export default ExportRequest;
