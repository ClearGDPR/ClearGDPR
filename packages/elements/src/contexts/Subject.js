import _ from 'lodash';
import { DATA_ERASED, DATA_STATUS } from '../constants';

export default class Subject {
  constructor(
    cg,
    {
      isFetched = false,
      isLoading = false,
      isErased = false,
      data = null,
      status = null,
      processors = null,
      objection = null,
      shares = null,
      restrictions = null,
      isGuest = true,
      propagateMutation
    }
  ) {
    this.cg = cg;

    this.isFetched = isFetched;
    this.isLoading = isLoading;
    this.isErased = isErased;
    this.data = data;
    this.status = status;
    this.processors = processors;
    this.objection = objection;
    this.shares = shares;
    this.restrictions = restrictions;
    this.isGuest = isGuest;

    this._propagateMutation = () => {
      propagateMutation(new Subject(cg, { ...this, propagateMutation }));
    };
  }

  async fetchData() {
    const { isLoading, isFetched, isGuest } = this;

    if (isLoading || isFetched || isGuest) {
      return;
    }

    this.isLoading = true;

    let data;
    let status;

    try {
      status = await this.cg.Subject.getDataStatus();
    } catch (e) {
      status = null;
    }

    const isErased = !status || status.controller === DATA_ERASED;

    if (!isErased) {
      try {
        data = await this.cg.Subject.accessData();
      } catch (e) {
        data = null;
      }
    }

    this.data = data;
    this.isLoading = false;
    this.isFetched = true;
    this.isErased = isErased;
    this.status = status;

    this._propagateMutation();
  }

  async fetchProcessors() {
    if (this.processors) {
      return;
    }

    let processors;
    try {
      processors = await this.cg.Subject.getProcessors();
    } catch (e) {
      console.log('failure', e);
      return;
    }

    this.processors = processors.map(p => {
      p.enabled = true;

      if (this.status) {
        const s = _.find(this.status.processors, { id: p.id }) || {};
        p.status = s.status ? DATA_STATUS[s.status] : DATA_STATUS.UNCONSENTED;
      }

      return p;
    });

    this._propagateMutation();
  }

  async initiateRectification(requestReason, rectificationPayload) {
    return await this.cg.Subject.initiateRectification(requestReason, rectificationPayload);
  }

  async eraseData() {
    await this.cg.Subject.eraseData();
    this.status.controller = DATA_ERASED;
    this.data = null;
    this.isErased = true;
    this._propagateMutation();
  }

  async fetchObjectionStatus() {
    if (this.objection !== null) {
      return;
    }
    const { objection } = await this.cg.Subject.getObjectionStatus();
    this.objection = objection;
    this._propagateMutation();
  }

  async updateObjection(value) {
    await this.cg.Subject.updateObjection(value);
    this.objection = value;
    this._propagateMutation();
  }

  async fetchDataShares() {
    if (this.shares !== null) {
      return;
    }
    try {
      this.shares = await this.cg.Subject.getDataShares();
    } catch (e) {
      console.log('failure', e);
      return;
    }

    this._propagateMutation();
  }

  async addDataShare(name) {
    try {
      await this.cg.Subject.addDataShare({
        name
      });
    } catch (e) {
      console.log(e);
      return;
    }
    this.shares = null;
    this.fetchDataShares();
  }

  async removeDataShare(id) {
    try {
      await this.cg.Subject.removeDataShare(id);
    } catch (e) {
      console.log(e);
      return;
    }
    this.shares = null;
    this.fetchDataShares();
  }

  async fetchRestrictions() {
    if (this.restrictions !== null) {
      return;
    }
    this.restrictions = await this.cg.Subject.getRestrictions();
    this._propagateMutation();
  }

  async updateRestrictions(restrictions) {
    this.restrictions = restrictions;
    this._propagateMutation();

    try {
      await this.cg.Subject.updateRestrictions(restrictions);
    } catch (e) {
      console.log(e);
    }
  }

  async giveConsent(payload) {
    try {
      await this.cg.Subject.giveConsent(payload);
    } catch (e) {
      console.log('failure', e);
    }
  }

  onEntrance(callback) {
    this.cg.Events.subscribe('auth.setAccessToken', callback);
  }

  clearEntranceListeners() {
    this.cg.Events.clear('auth.setAccessToken');
  }
}
