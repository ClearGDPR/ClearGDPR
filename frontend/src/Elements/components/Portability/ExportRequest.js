import React from 'react';
import PropTypes from 'prop-types';
import FileSaver from 'file-saver';
import Subject from 'contexts/Subject';

class ExportRequest extends React.PureComponent {
  state = {
    processing: false,
    success: false
  };

  loadData() {
    this.props.subject.fetchData();
    this.setState({ processing: true });
    const intervalId = setInterval(() => {
      const { data } = this.props.subject;
      if (!data) {
        return;
      }

      clearInterval(intervalId);

      const blob = new Blob([JSON.stringify(data)], { type: 'text/plain;charset=utf-8' });
      FileSaver.saveAs(blob, 'personal_data.json');
      this.setState({
        success: true,
        processing: false
      });
    }, 500);
  }

  render() {
    const { processing } = this.state;

    return (
      <button className="button is-primary" onClick={() => this.loadData()} disabled={processing}>
        {!processing ? (
          <React.Fragment>{this.props.options.label}</React.Fragment>
        ) : (
          <React.Fragment>Downloading...</React.Fragment>
        )}
      </button>
    );
  }
}

ExportRequest.propTypes = {
  options: PropTypes.object,
  subject: PropTypes.instanceOf(Subject)
};

export default ExportRequest;
