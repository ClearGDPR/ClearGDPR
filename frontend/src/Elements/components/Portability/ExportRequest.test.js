import React from 'react';
import { shallow } from 'enzyme';
import FileSaver from 'file-saver';
import ExportRequest from './ExportRequest';
import Subject from 'contexts/Subject';

jest.useFakeTimers();

const testData = {
  name: 'John',
  email: 'john@gmail.com'
};
const cg = {
  Subject: {
    accessData: async () => {
      return testData;
    }
  }
};

const subject = new Subject(cg, { data: testData });

const setup = () => {
  return shallow(<ExportRequest options={{ label: 'Load data' }} subject={subject} />);
};

describe('ExportRequest', () => {
  let spy;
  let saveAs;

  beforeEach(() => {
    saveAs = FileSaver.saveAs;
    FileSaver.saveAs = () => {};
  });

  it('should render correctly', () => {
    const component = setup();
    expect(component).toMatchSnapshot();
  });

  it('should toggle itself into processing state before export', () => {
    const component = setup();
    component.simulate('click');
    const { success, processing } = component.state();
    expect(success).toBe(false);
    expect(processing).toBe(true);
  });

  it('should request access to subjects data', async () => {
    spy = jest.spyOn(subject, 'fetchData');
    const component = setup();
    await component.simulate('click');
    expect(spy).toBeCalled();
  });

  it('should use FileSaver to save received data', async () => {
    spy = jest.spyOn(FileSaver, 'saveAs');
    const component = setup();
    await component.simulate('click');
    jest.runAllTimers();
    expect(spy).toBeCalledWith(expect.any(Blob), 'personal_data.json');
  });

  it('should toggle itself into success state', async () => {
    const component = setup();
    await component.simulate('click');
    jest.runAllTimers();
    const { success, processing } = component.state();
    expect(success).toBe(true);
    expect(processing).toBe(false);
  });

  afterEach(() => {
    if (spy) {
      spy.mockClear();
    }
    FileSaver.saveAs = saveAs;
  });
});
