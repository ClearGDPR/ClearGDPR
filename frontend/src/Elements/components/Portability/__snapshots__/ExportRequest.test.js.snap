// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportRequest should render correctly 1`] = `
<button
  className="button is-primary"
  disabled={false}
  onClick={[Function]}
>
  <React.Fragment>
    Load data
  </React.Fragment>
</button>
`;
